File: rabbitrack-mobile/lib/src/features/reports/litter_performance_report_controller.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../shared/offline_demo_data.dart';
import '../auth/auth_controller.dart';
import 'litter_performance_report_models.dart';
import 'litter_performance_report_repository.dart';

final litterPerformanceReportProvider =
    FutureProvider.autoDispose<LitterPerformanceReport>((ref) async {
      final session = ref.watch(authControllerProvider).valueOrNull;
      final farm = session?.selectedFarm;

      if (farm == null) {
        return _emptyLitterPerformanceReport;
      }

      if (isOfflineDemoSession(session)) {
        if (isOfflineEmptyFarm(farm.id)) {
          return _emptyLitterPerformanceReport;
        }

        final rows = offlineDemoLitters().map((litter) {
          final detail = offlineDemoLitterDetail(litter.id);
          final bornAlive = detail?.kitsBornAlive ?? litter.currentLiveCount;
          final stillborn = detail?.kitsStillborn ?? 0;
          final mortality = (bornAlive - litter.currentLiveCount).clamp(
            0,
            bornAlive,
          );

          return LitterPerformanceRow(
            id: litter.id,
            identifier: litter.identifier,
            doeIdentifier: litter.doeIdentifier,
            buckIdentifier: litter.buckIdentifier,
            kindledOn: litter.kindledOn,
            bornAlive: bornAlive,
            stillborn: stillborn,
            mortality: mortality,
            currentLive: litter.currentLiveCount,
            weaned: litter.status == 'weaned' ? litter.currentLiveCount : 0,
            survivalRate: bornAlive == 0
                ? 0
                : (litter.currentLiveCount / bornAlive) * 100,
            birthAverageWeight: detail?.weights.isEmpty == false
                ? detail!.weights.first.averageWeightValue
                : null,
            weightUnit: 'kg',
            status: litter.status,
          );
        }).toList();

        final bornAlive = rows.fold<int>(0, (sum, row) => sum + row.bornAlive);
        final currentLive = rows.fold<int>(
          0,
          (sum, row) => sum + row.currentLive,
        );
        final stillborn = rows.fold<int>(0, (sum, row) => sum + row.stillborn);
        final mortality = rows.fold<int>(0, (sum, row) => sum + row.mortality);
        final weaned = rows.fold<int>(0, (sum, row) => sum + row.weaned);

        return LitterPerformanceReport(
          litterCount: rows.length,
          bornAlive: bornAlive,
          stillborn: stillborn,
          mortality: mortality,
          currentLive: currentLive,
          weaned: weaned,
          survivalRate: bornAlive == 0 ? 0 : (currentLive / bornAlive) * 100,
          litters: rows,
        );
      }

      return ref.watch(litterPerformanceReportRepositoryProvider).show(farm.id);
    });

const _emptyLitterPerformanceReport = LitterPerformanceReport(
  litterCount: 0,
  bornAlive: 0,
  stillborn: 0,
  mortality: 0,
  currentLive: 0,
  weaned: 0,
  survivalRate: 0,
  litters: [],
);

File: rabbitrack-api/app/Http/Controllers/Api/V1/LitterPerformanceReportController.php
<?php

namespace App\Http\Controllers\Api\V1;

use App\Http\Controllers\Controller;
use App\Models\Farm;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Http\Response;

class LitterPerformanceReportController extends Controller
{
    public function show(Request $request, Farm $farm): JsonResponse|Response
    {
        $this->authorizeFarmAccess($request, $farm);

        $litters = $farm->litters()
            ->with(['doe', 'buck', 'weanings', 'weightRecords'])
            ->orderByDesc('kindled_on')
            ->get();

        $bornAlive = (int) $litters->sum('kits_born_alive');
        $stillborn = (int) $litters->sum('kits_stillborn');
        $currentLive = (int) $litters->sum('current_live_count');
        $weaned = (int) $litters->sum(fn ($litter) => $litter->weanings->sum('number_weaned'));
        $mortality = max(0, $bornAlive - $currentLive);

        $data = [
            'litter_count' => $litters->count(),
            'born_alive' => $bornAlive,
            'stillborn' => $stillborn,
            'mortality' => $mortality,
            'current_live' => $currentLive,
            'weaned' => $weaned,
            'survival_rate' => $bornAlive > 0 ? round(($currentLive / $bornAlive) * 100, 1) : 0.0,
            'litters' => $litters->map(fn ($litter) => $this->litterRow($litter))->values(),
        ];

        if ($request->query('format') === 'csv') {
            return $this->csv($data);
        }

        return response()->json(['data' => $data]);
    }

    private function litterRow($litter): array
    {
        $birthWeight = $litter->weightRecords->firstWhere('stage', 'birth');
        $weaningWeight = $litter->weightRecords->firstWhere('stage', 'weaning');
        $weaned = (int) $litter->weanings->sum('number_weaned');
        $mortality = max(0, $litter->kits_born_alive - $litter->current_live_count);

        return [
            'id' => $litter->id,
            'identifier' => $litter->identifier,
            'doe_identifier' => $litter->doe?->identifier,
            'buck_identifier' => $litter->buck?->identifier,
            'kindled_on' => $litter->kindled_on?->toDateString(),
            'born_alive' => $litter->kits_born_alive,
            'stillborn' => $litter->kits_stillborn,
            'mortality' => $mortality,
            'current_live' => $litter->current_live_count,
            'weaned' => $weaned,
            'survival_rate' => $litter->kits_born_alive > 0
                ? round(($litter->current_live_count / $litter->kits_born_alive) * 100, 1)
                : 0.0,
            'birth_average_weight' => $birthWeight?->average_weight_value,
            'weaning_average_weight' => $weaningWeight?->average_weight_value,
            'weight_unit' => $weaningWeight?->weight_unit ?? $birthWeight?->weight_unit ?? 'kg',
            'status' => $litter->status,
        ];
    }

    private function csv(array $data): Response
    {
        $handle = fopen('php://temp', 'r+');
        fputcsv($handle, [
            'identifier',
            'doe_identifier',
            'buck_identifier',
            'kindled_on',
            'born_alive',
            'stillborn',
            'mortality',
            'current_live',
            'weaned',
            'survival_rate',
            'birth_average_weight',
            'weaning_average_weight',
            'weight_unit',
            'status',
        ]);

        foreach ($data['litters'] as $litter) {
            fputcsv($handle, [
                $litter['identifier'],
                $litter['doe_identifier'],
                $litter['buck_identifier'],
                $litter['kindled_on'],
                $litter['born_alive'],
                $litter['stillborn'],
                $litter['mortality'],
                $litter['current_live'],
                $litter['weaned'],
                $litter['survival_rate'],
                $litter['birth_average_weight'],
                $litter['weaning_average_weight'],
                $litter['weight_unit'],
                $litter['status'],
            ]);
        }

        rewind($handle);
        $content = stream_get_contents($handle);
        fclose($handle);

        return response($content, 200, [
            'Content-Type' => 'text/csv',
            'Content-Disposition' => 'attachment; filename="litter-performance-report.csv"',
        ]);
    }

    private function authorizeFarmAccess(Request $request, Farm $farm): void
    {
        $hasAccess = $request->user()
            ->memberships()
            ->where('farm_id', $farm->id)
            ->where('is_active', true)
            ->exists();

        abort_unless($hasAccess, 404);
    }
}
